### PR TITLE
PluginController: Eligible Species

### PR DESCRIPTION
**Pull request by ax3l**
Only specialize and compile particle plugins for particle species which are elibible for it, based on the `SpeciesEligibleForSolver` trait.

Silently disables components which are not fulfilled, e.g. for nonphysical particles such as probes. Moves user (input) errors from compile time (which are in this case always very long and unreadable template meta programming errors) to runtime where it is (ideally) easier to spot.

Possible follow-up: For further usability, we can add log messages on high-level debug levels to output which species are ignored on `pluginLoad()`.

Depends on

- [x] #2364
- [x] #2367
- [x]  #2380 
- [x] #2394
- [x] CT: no significant impact after 2nd commit

## Compile Times LWFA

**CUDA 7.5**
`dev`: 6:03 (real) | 7:38 (user)
`PR`: 6:08m (real) | 7:40m (user)

**CUDA 8.0.44** + ISAAC
`dev`: 6:35m (real) | 7:36m (user)
`PR`: 6:46m (real) | 7:43m (user)

### CODE REVIEW

File: include/picongpu/plugins/PluginController.hpp [hunk 1]
```diff
@@ -115,8 +117,40 @@ class PluginController : public ILightweightPlugin
     {
     };
 
+    /** Return a list of Plugins specialized to all matching species
+     *
+     * Plugins can define the trait SpeciesAllowedWithSolver to filter a
```
[the remark this answers is not in the document]
Comment (ax3l) [reply include/picongpu/plugins/PluginController.hpp]: outdated comment, you are right :)

File: include/picongpu/particles/traits/ApplyEligibleSpecies.hpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+     *
+     * @tparam T_Solver a particle solver which shall be specialized for all
+     *                  eligible particle species
+     * @tparam T_SeqSpecies a sequence of particle species to check if they are
+     *                      eligible to specialize T_Solver, also allows a
+     *                      single type instead of a sequence
+     */
+    template<
+        typename T_Solver,
+        typename T_SeqSpecies
```
Comment (ax3l) [inline include/picongpu/particles/traits/ApplyEligibleSpecies.hpp]: this signature will get an update

File: include/picongpu/particles/traits/ApplyEligibleSpecies.hpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    >
+    struct ApplyEligibleSpecies
+    {
+        // wrap single arguments to sequence
+        using SeqSpecies = typename pmacc::ToSeq< T_SeqSpecies >::type;
+        // unspecialized solver
+        using Solver = T_Solver;
+
+        template< typename T_Species >
+        struct Op : bmpl::apply1<
```
Comment (ax3l) [inline include/picongpu/particles/traits/ApplyEligibleSpecies.hpp]: might need `::type` here and change the usage

**Comment by psychocoderHPC** [comment]
can rebased against #2380

**Comment by ax3l** [comment]
this PR is causing `cudafe(++)` to hand with CUDA 7.5 on the compile suite.

File: include/picongpu/plugins/PluginController.hpp [hunk 4]
```diff
@@ -183,25 +188,25 @@ class PluginController : public ILightweightPlugin
...
-
-    typedef bmpl::transform<
-        CombinedUnspecializedSpeciesPlugins,
-        ApplyDataToPlugin<bmpl::_1>
-    >::type SpeciesPlugins;
-
-
-    /* create sequence with all plugins*/
-    typedef MakeSeq<
+    using SpeciesPlugins = typename MakeSeqFromNestedSeq<
```
Comment (ax3l) [inline include/picongpu/plugins/PluginController.hpp]: removed this section now to use `SpeciesEligibleForSolver` directly with no noticable CT impact.

One reason for the overhead could also be the `MakeSeqFromNestedSeq`.

**Pull request merged**